### ISSUE TITLE
Optimize 2D-CNN

### ISSUE THREAD

**Issue by ValleSoYeah** [tamap94/ds-capstone-alzheimers-#43]
- [x] implement tracking of model performances

**Comment by ValleSoYeah** [comment]
start with one image slice and when improve the model also try different slices

**Comment by deneick** [comment]
- [ ] Different slices from one patient for training
- [ ] Rescaling values
- [ ] Removing young patients
- [ ] Balancing classes 
- [ ] Use skull removed 

- [ ] Majority voting for different views 
- [ ] Cyclic learning rate
- [ ] Batch normalization 
- [ ] Using pca to get eigenbrains
- [ ] Resnet50 or vgg16 with random weights
- [ ] Transfer learning 
- [ ] Attention layers

**Comment by ValleSoYeah** [comment]
Recent sucesses:

- Locally connected conv2d improves validation accuracy and reduces overfitting even in very simple model
- 98% validation accuracy with pretrained VGG16 + LocConnectedConv2D layer (without dropping data and loading multiple slices for training)
- still only 73% test accuracy in best model => only single slices used for X_test (mimics application of the model to predict from 2D images alone)

**Comment by ValleSoYeah** [comment]
Try next:
- [ ] Shift each ADNI brain so its mean aligns with the mean of all OASIS brains
- [ ] use VGG16_LocConn model to identify best slices
- [ ] modify get_slice functions to allow loading of thin 3D images (3 layers, but connected like in an RGB image)
- [ ] can eigenbrains increase training time? do CNNs work with eigenbrains?

**Comment by deneick** [comment]
- I tried using pca with svm with 2d and 3d.
Problem: we can only have as many components as data samples. 

- I changed the loading of the data, so it loads the data as uint8 and int16 instead of converting to float64
Makes the loading of the data faster and needs much less memory, but when applying pca or cnn, the data needs to be converted to float32 or float64
Data also gets converted when normalizing in current code

- data can also get sliced before even loading the array. Could make loading slices faster
[nibabel doc](https://nipy.org/nibabel/images_and_memory.html)

- cnns can be built with multiple inputs. Can be used for different views 
[code](https://www.kaggle.com/code/hireme/two-inputs-neural-network-using-keras/notebook)
